showroom/comments.py

```diff
@@ -57,80 +188,275 @@ def __init__(self, room, client, settings, watcher):
+                cmt_logger.error('no JSON message - {}'.format(message))
+                return
+            message = message[idx:]
+            try:
+                data = json.loads(message)
+            except JSONDecodeError as e:
+                cmt_logger.debug('broken message, JSON decode error: {}'.format(e))
+                cmt_logger.debug('--> {}'.format(message))
+                # try to fix
+                message += '","t":"1"}'
+                try:
+                    data = json.loads(message)
+                except JSONDecodeError:
+                    cmt_logger.error('failed to fix broken message, JSON decode error: {}'.format(message))
+                    return
+                cmt_logger.debug('--> fix passed: {}'.format(message))
+
+            # add current time
+            data['received_at'] = now
+
+            # Some useful info in the message:
+            # ['t']  message type, determine the message is comment, telop, or gift
+            # ['cm'] comment
+            # ['ac'] name
+            # ['u']  user_id
+            # ['av'] avatar_id
+            # ['g'] gift_id
+            # ['n'] gift_num
+
+            # type of the message
+            m_type = str(data['t'])  # could be integer or string
+
+            if m_type == '1':  # comment
+                comment = data['cm']
+
+                # skip counting for 50
+                if len(comment) < 3 and comment.isdecimal() and int(comment) <= 50:
+                    # s1 = '⑷'; s2 = u'²'; s3 = '❹'
+                    # print(s1.isdigit())  # True
+                    # print(s2.isdigit())  # True
+                    # print(s1.isdecimal())  # False
+                    # print(s2.isdecimal())  # False
+                    # int(s1)  # ValueError
+                    # int(s2)  # ValueError
+                    pass
+                else:
+                    comment = comment.replace('\n', ' ')  # replace line break to a space
+                    # cmt_logger.info('{}: {}'.format(self.room.name, comment))
+                    data['cm'] = comment
+                    self.comment_log.append(data)
+                    self.comment_count += 1
+
+            elif m_type == '2':  # gift
+                pass
+
+            elif m_type == '8':  # telop
+                self.comment_log.append(data)
+                if data['telop'] is not None:  # could be null
+                    # cmt_logger.info('{}: telop = {}'.format(self.room.name, data['telop']))
+                    pass
+
+            elif m_type == '11':  # cumulated gifts report
+                pass
+
+            elif m_type == '101':  # indicating live finished
+                self.comment_log.append(data)
+                self._isQuit = True
+
+            else:
+                self.comment_log.append(data)
+
+        def ws_on_error(ws, error):
+            """ WebSocket callback """
+            cmt_logger.error('websocket on error: {}'.format(error))
```

Comment:
Fixing the encoding errors might be as simple as changing this to:

```python
        def ws_on_error(ws, error):
            """ WebSocket callback """
            if isinstance(error, UnicodeDecodeError):
                data = error.object
                try:
                    data = data.decode('latin-1')
                except UnicodeDecodeError:
                    pass
                else:
                    ws_on_message(ws, data)
                    return
            cmt_logger.error('websocket on error: {}'.format(error))
```

Reply:
Thank you for the explanation and the code. I see now we can have the correct encoded message. I have a question that when this method returns, the websocket will still stop like before, right? I will make the change but it will take time for me to test it. I cannot record too many rooms on my computer so actually I rarely met this error.

Reply:
If you know what room encountered this error a lot, please let me know.